---
title: "GSE80080"
output: html_document
date: "2023-09-15"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Analysis of microarray data from GSE80080

```{r}
library(limma)
library(knitr)
library("illuminaio")

bgxfile = dir(path="./", pattern="bgx")
idatfiles = dir(path="./GSE80080_RAW", pattern=".idat", full.names = T)
# idatfiles = idatfiles[-4] # removing RXF2
data = read.idat(idatfiles, bgxfile)
data$other$Detection <- detectionPValues(data)

```




```{r}
pe = propexpr(data)
kable(pe, "markdown") # proprotion of probes being detected as expressed, should be consistent
```




```{r}
data.norm = neqc(data) # normalize data
# table(apply(data.norm$other$Detection, 1, function(x){all(x < 0.05)}))

data.norm

```


```{r}
sampledata = data.frame(geo=c("GSM2112539", "GSM2112540", "GSM2112541", "GSM2112542", "GSM2112543", "GSM2112544"),
						name = c("SKR_1", "RXF_1", "SKR_2", "RXF_2", "SKR_3", "RXF_3"),
						condition = c("SKR", "RXF", "SKR", "RXF", "SKR", "RXF")
						)

#without RXF 2
# sampledata = data.frame(geo=c("GSM2112539", "GSM2112540", "GSM2112541", "GSM2112543", "GSM2112544"),
# 						name = c("SKR_1", "RXF_1", "SKR_2", "SKR_3", "RXF_3"),
# 						condition = c("SKR", "RXF", "SKR", "SKR", "RXF")
# 						)
```


```{r}
eset = data.norm$E
distance = dist(t(eset))#,method="maximum")
clusters = hclust(distance, method="ward.D")
for (i in seq(length(clusters$labels))){
  if (clusters$labels[i] == "./GSE80080_RAW/GSM2112539_9533701055_A_Grn") {
    clusters$labels[i] = "SKR_1"
  } else if (clusters$labels[i] == "./GSE80080_RAW/GSM2112540_9533701055_B_Grn") {
    clusters$labels[i] = "RXF_1"
  } else if (clusters$labels[i] == "./GSE80080_RAW/GSM2112541_9533701055_E_Grn") {
    clusters$labels[i] = "SKR_2"
  } else if (clusters$labels[i] == "./GSE80080_RAW/GSM2112542_9533701055_F_Grn") {
    clusters$labels[i] = "RXF_2"
  } else if (clusters$labels[i] == "./GSE80080_RAW/GSM2112543_9533701055_I_Grn") {
    clusters$labels[i] = "SKR_3"
  } else {
    clusters$labels[i] = "RXF_3"
  }
}
plot(clusters, main = NULL, xlab = NULL) 
```





```{r}
library(ggplot2)

pca = prcomp(t(eset))
percentVar <- round(100*pca$sdev^2/sum(pca$sdev^2))
data = as.data.frame(pca$x)
data$name = sampledata$name
data$condition = sampledata$condition
ggplot(data, aes(PC1, PC2, colour=condition)) + geom_point(size=4) +
  scale_x_continuous(paste0("PC1: ",percentVar[1],"% variance"), limits=c(-45, 45))  + 
  scale_y_continuous(paste0("PC2: ",percentVar[2],"% variance"), limits=c(-45, 45)) +
  coord_fixed()  + geom_text(data = data, aes(PC1,PC2, label = name), hjust = 1.2) + 
  scale_color_manual(values = c("firebrick", "navy"),
                     labels = c("Cachectic", "Control"),
                     name = "Condition")
```


```{r}
condition <- factor(sampledata$condition, unique(sampledata$condition))
design <- model.matrix(~0 + condition)
colnames(design) <- c("SKR", "RXF")
head(design) # design matrix: ???
```



```{r without unexpressed probes}

# table(apply(data.norm$other$Detection, 1, function(x){any(x < 0.05)})) # how many are expressed

# data.norm.filtered = data.norm[apply(data.norm$other$Detection, 1, function(x){any(x < 0.05)}),] # expressed in any sample
data.norm.filtered = data.norm[(rowSums(data.norm$other$Detection < 0.05) >= 3),] # expressed in at least 3 samples

```

```{r without the probes without genes}
table(table(data.norm$genes$Symbol))

 
empty_probes = c(which(data.norm.filtered$genes$Symbol == ""), which(data.norm.filtered$genes$Symbol == " "), which(is.na(data.norm.filtered$genes$Symbol)))
data.norm.filtered$genes = data.norm.filtered$genes[-empty_probes,]
nrow(data.norm.filtered$genes)

data.norm.filtered$E = data.norm.filtered$E[rownames(data.norm.filtered$E) %in% data.norm.filtered$genes$Array_Address_Id,]
nrow(data.norm.filtered$E) # 18533

# remove_empty_probes <- function(E, genes){
#   
#   print(dim(E))
#   E_remove = c()
#   for (i in seq(length(rownames(E)))){
#     if (!(rownames(E)[i] %in% genes$Array_Address_Id)){
#       E_remove = c(E_remove, i)
#     }
#   }
#   E = E[-E_remove,]
#   print(dim(E))
#   return(E)
# }
# 
# data.norm.filtered$E = remove_empty_probes(data.norm.filtered$E, data.norm.filtered$genes)
# nrow(data.norm.filtered) # 18533




```

```{r without redundant probes}

#n
# empty_probes = c(which(data.norm.filtered$genes$Symbol == ""), which(data.norm.filtered$genes$Symbol == " "), which(is.na(data.norm.filtered$genes$Symbol)))
# x = data.norm.filtered$genes[-empty_probes,]

widest_range_access = c()
all_repeat_access_ids = c() # all access IDs associated with repeated genes 
for (i in unique(data.norm.filtered$genes$Symbol[which(duplicated(data.norm.filtered$genes$Symbol))])) {
  quartile_range = c()
  access_id = as.character(data.norm.filtered$genes$Array_Address_Id[data.norm.filtered$genes$Symbol == i]) # collect access IDs of those with multiple gene symbols
  all_repeat_access_ids = c(all_repeat_access_ids, access_id)
  for (t in access_id){ # for every probe, calculate the interquartile range
    quartile_range = c(quartile_range, abs(summary(data.norm.filtered$E[t,])[5] - summary(data.norm.filtered$E[t,])[2]))
  }
  widest_range_access = c(widest_range_access, access_id[which.max(quartile_range)])
}

remove_repeats = all_repeat_access_ids[!(all_repeat_access_ids %in% widest_range_access)]
data.norm.filtered = data.norm.filtered[-(which(rownames(data.norm.filtered) %in% remove_repeats)),]
nrow(data.norm.filtered) #14428

# number of unique genes detected
length(unique(data.norm.filtered$genes$Symbol[data.norm.filtered$genes$Array_Address_Id %in% rownames(data.norm.filtered$E)]))
length(unique(data.norm.filtered$genes$Symbol)) # 14428 genes




```

```{r}
# DE 

fit <- lmFit(data.norm.filtered, design)
contrast.matrix <- makeContrasts(kc_knc = RXF - SKR, levels=design)
kc_fits <- contrasts.fit(fit, contrast.matrix)
kc_ebFit <- eBayes(kc_fits)
results = topTable(kc_ebFit, coef=1, number=nrow(data.norm.filtered))
sum(results$P.Value < 0.01) 
# 1196 genes, not accounting for multiple testing (with: all samples, expressed genes at (any < 0.05), no empty probes and repeat probes with lower interquartile range)
# *1527 genes, not accounting for multiple testing (with: all samples, expressed genes in at least 3 samples. no empty probes and repeat probes with lower interquartile range)
# 800 genes, not accounting for multiple testing (with: excluding RXF2, expressed genes at (any < 0.05), no empty probes and repeat probes with lower interquartile range)
# 1079 genes, not accounting for multiple testing (with: excluding RXF2, expressed genes in at least 3 samples. no empty probes and repeat probes with lower interquartile range)



sum(results$P.Value < 0.1) 
# 5102 genes, not accounting for multiple testing (with: all samples, expressed genes at (any < 0.05), no empty probes and repeat probes with lower interquartile range)
# *4930 genes, not accounting for multiple testing (with: all samples, expressed genes in at least 3 samples. no empty probes and repeat probes with lower interquartile range)
# 4401 genes, not accounting for multiple testing (with: excluding RXF2, expressed genes at (any < 0.05), no empty probes and repeat probes with lower interquartile range)
# 4308 genes, not accounting for multiple testing (with: excluding RXF2, expressed genes in at least 3 samples. no empty probes and repeat probes with lower interquartile range)


sum(results$adj.P.Val < 0.1) 
# 299 with: all samples, expressed genes at (any < 0.05), no empty probes and repeat probes with lower interquartile range
# *1655 with: all samples, expressed genes in at least 3 samples, no empty probes and repeat probes with lower interquartile range -- YAY
# 0! with: without RXF2, expressed genes at (any < 0.05), no empty probes and repeat probes with lower interquartile range
# 577 with: all samples, expressed genes in at least 3 samples, no empty probes and repeat probes with lower interquartile range

de_results = results[results$adj.P.Val < 0.1,]
de_up_results = de_results[de_results$logFC > 0,] # upregulated genes
de_down_results = de_results[de_results$logFC < 0,] # downregulated genes

sum(de_results$logFC > 0)
hist(de_results$logFC)


hist(results$P.Value, xlab="pvalue", main="") 
hist(results$adj.P.Val, xlab="pvalue", main="") 


d = as.data.frame(results)
d$color = ifelse(-log10(results$adj.P.Val) > -log10(0.1) & results$logFC > 0,  "up", ifelse(-log10(results$adj.P.Val) > -log10(0.1) & results$logFC < 0,  "down", "not_sig"))
ggplot(data = d) + 
  geom_point(mapping = aes(x = logFC, y = -log10(adj.P.Val), color = color)) + 
  scale_x_continuous(name = "Log2FoldChange", limits = c(-1.6,1)) + 
  scale_y_continuous(name = "-Log10 Adjusted p-value") + 
  scale_color_manual(name = NULL, labels = c("Downregulated", "Not Significant", "Upregulated"),
                     values = c("up" = "firebrick", "down" = "navy", "not_sig" = "gray"))

# volcanoplot(kc_ebFit, col = ifelse(-log10(kc_ebFit$p.value) >= -log10(0.1) & kc_ebFit$coefficients > 0, "firebrick", ifelse(-log10(kc_ebFit$p.value) >= -log10(0.1) & kc_ebFit$coefficients < 0, "navy", "gray")))

limma::plotMA(kc_ebFit, main = NULL, col = ifelse(kc_ebFit$coefficients > 0, "firebrick", ifelse( kc_ebFit$coefficients < 0, "navy", "gray")))

```


```{r eval = FALSE}

results.export = results[,c(3,4,8)]
sum(results.export$adj.P.Val < 0.1 & results.export$logFC < 0) # 1001 downregulated genes (corresponding to upregulated miRNA)
sum(results.export$adj.P.Val < 0.1 & results.export$logFC > 0) # 654 upregulated genes (corresponding to downregulated miRNA)
results.export$change = ifelse(results.export$adj.P.Val < 0.1 
                                   & results.export$logFC > 0, "UPregulated",
                                   ifelse(results.export$adj.P.Val < 0.1 
                                   & results.export$logFC < 0, "DOWNregulated",
                                   "not significant"))
rownames(results.export) = seq(nrow(results.export))
write.csv(results.export, file = "all_microarray_results.csv") # all genes differentially expressed


write.csv(de_results, file = "de_genes_microarray.csv") # all genes differentially expressed

write.csv(data.norm.filtered$genes, file =  "expressed_genes_microarray.csv") # ALL expressed genes in the dataset, not necessarily differentially expressed
```


```{r GSEA of all expressed genes}

library(fgsea)

set.seed(42) ## TO DO !

pathways <- c(gmtPathways("h.all.v2023.1.Hs.symbols.gmt"), gmtPathways("c2.cp.v2023.1.Hs.symbols.gmt")) 


changes_lfc = results$logFC
names(changes_lfc) = results$Symbol

changes_lfc = changes_lfc[order(changes_lfc, decreasing=TRUE)] #ranked list
head(changes_lfc)
head(pathways)

fgseaRes <- fgsea(pathways = pathways, 
                  stats = changes_lfc,
                  minSize=15,
                  maxSize=500,
                  eps = 0.0) #to estimate p-values more accurately

head(fgseaRes[order(pval), ], n = 33)
sum(fgseaRes$padj < 0.1)
fgseaRes$pathway[fgseaRes$padj < 0.1 & fgseaRes$NES > 0] #"HALLMARK_PANCREAS_BETA_CELLS"
length(unique(fgseaRes$pathway[fgseaRes$padj < 0.1 & fgseaRes$NES < 0])) #32 down, 1 up

changing_pathways = fgseaRes[fgseaRes$padj < 0.1,]
changing_pathways$pathway

changing_pathways_for_export = changing_pathways[,-c(8)]

# export table
# write.table(changing_pathways_for_export, "GSEA_table.txt", sep = ",", 
#             quote = FALSE, row.names = F)

# fgseaRes <- fgsea(pathways = examplePathways, 
#                   stats    = exampleRanks,
#                   eps      = 0.0,
#                   minSize  = 15,
#                   maxSize  = 500)
```

# Pathway 1: Hallmark Epithelial Mesenchymal Transition 

EXPLANATION: 
"Epithelial-mesenchymal transition (EMT) is a biological program whereby epithelial cells acquire mesenchymal phenotype... cancer cells that undergo EMT, apart from invasive, metastatic properties, also exhibit stem cell traits and resistance to programmed cell death...EMT has been linked to tumor aggressiveness and worse clinical features, which is not surprising, since EMT promotes several malignant characteristics including migration and invasion, stemness, metabolic reprogramming, immune evasion and resistance to therapy." (https://link-springer-com.libproxy1.nus.edu.sg/article/10.1007/s11010-023-04670-x)

Cancer cells "lose epithelial characteristics and integrity, acquire mesenchymal features, and become motile" through EMT and fuel invasion of other parts of the body. "EMT is increasingly understood to orchestrate a large variety of complementary cancer features, such as tumor cell stemness, tumorigenicity, resistance to therapy and adaptation to changes in the microenvironment." (https://www-embopress-org.libproxy1.nus.edu.sg/doi/full/10.15252/embj.2021108647)

Notes: No information on how it is related to cachexia? maybe because cachexia occurs in more advanced cancers?
```{r}

plotEnrichment(pathways[["HALLMARK_PANCREAS_BETA_CELLS"]], changes_lfc, ticksSize = 1) + labs(title="HALLMARK_PANCREAS_BETA_CELLS")

```



# Pathway 2: Hallmark Hypoxia

EXPLANATION: 
"It is thus speculated that during cancer initiation and progression, [Hypoxia-inducible factor 1a] greatly helps pancreatic cells to gain oncogenic properties that allow them to face and overcome the hypoxic conditions, like high proliferative capacity, invasion and metastatic potential...tumor and stromal HIF-1α signaling enhances innate radiation resistance of hypoxic pancreatic tumor cells." (https://www.mdpi.com/2072-6694/2/4/2138)

Hypoxia could stimulate glutamine import, glutamine can be used to fuel proliferation of pancreatic cancer cells (along with glucose). (https://www.mdpi.com/2072-6694/2/4/2138)

"increased HIF-1α in the heart, EDL, soleus, and diaphragm of tumor-bearing mice...  The effect of hypoxia and subsequent increase in HIF-1α in the heart has both beneficial and detrimental effects. Transient hypoxia and induction of HIF-1α have been shown to be cardioprotective in high-altitude training and in myocardial infarction (20, 31, 51). A plausible mechanism for this is that HIF-1α regulates the oxygen consumption of mitochondria, providing protection during hypoxic events and decreased oxygen availability (36)." (https://journals.physiology.org/doi/full/10.1152/ajpheart.00090.2016)
```{r}

plotEnrichment(pathways[[changing_pathways$pathway[2]]], changes_lfc, ticksSize = 1) + labs(title=changing_pathways$pathway[2])

```

# Pathway 3: Hallmark TNFA Signaling via NFKB -- SAVE FOR FIGURES

EXPLANATION: 
NF-kB is a transcription factor that is activated in response to many things such as pro-inflammatory cytokine, TNFa. TNFa is produced by many cell types such as smooth muscle cells in skeletal muscle and adipocytes. (https://link.springer.com/chapter/10.1007/978-981-13-1435-3_12)

"TNF-α is responsible for the increase in gluconeogenesis, loss of adipose tissue and proteolysis, while causing decrease in protein, lipid and glycogen synthesis...TNF-α increases the corticotrophin-releasing hormone (CRH) level and decreases food intake...NF-kB induced catabolic signaling plays a key role in protein degradation associated with cachexia...TNF-α activates nuclear factor kappa B (NF-κB) which is a primary mediator for the control of transcription and a major applicant for signaling during catabolism. TNF-α rapidly promotes activation of NF-kB in cells of skeletal muscle, including differentiable myotubes and undifferentiable myoblasts...NF-kB affects on the expression of genes which regulate ubiquitin proteasome pathway (UPP) and undoubtedly promotes the loss of protein" (https://www-sciencedirect-com.libproxy1.nus.edu.sg/science/article/pii/S0024320516306853#bb0170)

TNF-a leads to spikes in NF-kB activity leading to prolonged changes in protein levels in muscles (protein loss). (https://respiratory-research.biomedcentral.com/articles/10.1186/rr67)

"NF-κB was activated in muscle progenitors during cachexia, leading to sustained Pax7 expression. In turn, a self-renewing signal in myogenic cells is likely propagated that results in their inability to progress through a differentiation program...activation of NF-κB leading to Pax7 appears crucial in directing proper muscle differentiation at different stages of development and under different environmental conditions." (https://www-ncbi-nlm-nih-gov.libproxy1.nus.edu.sg/pmc/articles/PMC3809785/)
```{r}

plotEnrichment(pathways[[changing_pathways$pathway[7]]], changes_lfc, ticksSize = 1) + labs(title=changing_pathways$pathway[7])

```

# Pathway 4: KEGG Pathways in Cancer

EXPLANATION: 
PI3K, one of the pathways in KEGG Pathways in Cancer, activates mTORC1 signalling whose "reduction ... in skeletal muscle contributes to loss of muscle mass during cachexia" (https://www.mdpi.com/1422-0067/19/8/2225)

"These indicated that selumetinib effectively prevented skeletal muscle wasting in cancer cachexia model through ERK inhibition and AKT activation in gastrocnemius muscle via cross-inhibition... Activation of MEK/ERK pathway was detected in the wasted muscle tissue of in vivo cancer cachexia model and in the differentiated myotubes of in vitro myofiber atrophy model... MEK/ERK pathway activates ubiquitination during muscle proteolysis. MEK phosphorylates ERK and activates the cascade of cytoplasmic signaling proteins and transcription factors, which regulate the genes related to ubiquitin proteasome. Selumetinib treatment resulted in decreased phosphorylation of ERK1/2 (Fig. 7). Activation of MEK/ERK pathway enhanced ubiquitination. Therefore, the decrease in E3 ligases MuRF1 and Fbx32, rather than excessive production of proinflammatory cytokines, followed selumetinib treatment in the cachexia group." (https://aacrjournals.org/mct/article/16/2/334/92273/Selumetinib-Attenuates-Skeletal-Muscle-Wasting-in)
```{r}

plotEnrichment(pathways[[changing_pathways$pathway[4]]], changes_lfc, ticksSize = 1) + labs(title=changing_pathways$pathway[4])

```


# Pathway 5: NABA Matrisome

EXPLANATION: 
Pathway was also enriched in another study on cachectic patients (https://www.frontiersin.org/journals/oncology/articles/10.3389/fonc.2022.1040112/full)

"TGFβ1 and TGFβ3[, ECM glyocoproteins,] levels were up-regulated by cachexia in AT, as well in the isolated adipocytes... The TGFβ signaling pathway can directly induce myofibroblasts to produce ECM components... We report that all of the studied collagen types (COL1A1, COL3A1, and COL6A1) in the AT, to be modulated by cancer cachexia...In fact, uncontrolled deposition of collagens, and of other ECM proteins, can prevent the expansion of adipocytes thus leading to the failure of treatments aiming at recovering adipose tissue wasting in cachectic patients." (https://link.springer.com/article/10.1186/s12885-017-3178-8) 
```{r}

plotEnrichment(pathways[[changing_pathways$pathway[12]]], changes_lfc, ticksSize = 1) + labs(title=changing_pathways$pathway[12])

```

# Pathway 6: NABA Matrisome Associated

EXPLANATION: 
Cardiac and skeletal muscles in CAC mouse model show upregulation of certain Matrix metalloproteinases, enzymes that degrade ECM. "cancer cachexia-associated matrix remodeling results in cardiac fibrosis and possible skeletal muscle remodeling... MMPs are actively remodeling the ECM and increasing collagen deposition. This MMP activity and matrix remodeling are most likely responsible for the decreases in wall thickness during contraction, contributing to decreased cardiac function... This aberrant increase in MMPs in the EDL and soleus muscles in the cachectic mouse model may contribute to the inability of satellite cell proliferation by breaking down the surrounding matrix and disrupting normal cell survival signals transmitted through the matrix and between cells." (https://journals.physiology.org/doi/full/10.1152/ajpheart.00106.2015) 
```{r}

plotEnrichment(pathways[[changing_pathways$pathway[6]]], changes_lfc, ticksSize = 1) + labs(title=changing_pathways$pathway[6])

```

# Pathway 7: PID FRA Pathway

EXPLANATION: 
"accumulating studies support that increased expression of AP-1 is associated with a variety of human cancers, and AP-1 is actively implicated in tumor progression and development." (https://pubs.acs.org/doi/full/10.1021/jm5004733#)

"After activation, AP-1 regulates the expression of downstream genes involved in various aspects of cancer biology, such as cell growth, apoptosis, angiogenesis, invasion, metastasis and drug resistance." (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10361657/)

Increased Fra-1 in C2C12 mouse muscle cells comes with decreased myosin heavy chain (MyHC) expression. "When we overexpress the Fra-1 subunit in C2C12 cells, we see increased mRNA expression of FAS and caspases 1, 8 and 11 (ortholog of human caspase 4). Caspase 11-mediated activation of caspase 1 appears to activate IL-1β, a catabolism-inducing cytokine implicated in cachexia.  Downstream of Fra-1 expression, we also see increased caspase 3 activity and an increase in the number of TUNEL-positive apoptotic cells... TNFα signaling through the MAPK pathway leads to phosphorylation-dependent stabilization of Fra-1, and leads to a switch to Jun:Fra-1 heterodimer composition which, in turn, transcriptionally up-regulates FAS and caspases 1, 8 and 11." This leads to increased apoptosis. (https://aacrjournals.org/cancerres/article/71/8_Supplement/4786/573579/Abstract-4786-A-role-for-increased-Fra-1)
```{r}

plotEnrichment(pathways[[changing_pathways$pathway[7]]], changes_lfc, ticksSize = 1) + labs(title=changing_pathways$pathway[7])

```

# Pathway 8: WP Adipogenesis

EXPLANATION: 
"Although skeletal muscle atrophy due to increased protein breakdown is known as the major hallmark of CAC, depletion and remodeling of adipose tissue also plays a crucial role in cachectic patients. The mechanism that drives adipose tissue depletion is multifactorial. Lipolysis is activated, accompanied by elevated total energy expenditure and suppressed adipogenesis, eventually reducing the fat mass. Several studies have revealed that adipose tissue wasting typically occurs before the appearance of other classic cachexia signs." (https://www.sciencedirect.com/science/article/pii/S2212877821001575)

Upregulation of adipogenesis inhibitors, coupled with increased lipolysis, leads to decreased fat mass?

"Besides, the extracellular matrix (ECM) plays a crucial role in adipocyte differentiation." --> NABA Matrisome??
"Adipose tissue remodeling is associated with several chronic diseases, including cancer cachexia. These changes comprise morphological and structural modifications characterized by adipocyte atrophy due to impairment of several processes as increased lipid mobilization, enhanced triglyceride lipolysis, reduced lipogenesis, impaired adipogenesis, extracellular matrix (ECM) rearrangement (generally resulting in fibrosis), enhanced inflammation, and adipose tissue “browning”." (https://www.mdpi.com/1422-0067/21/14/4860)
```{r}

plotEnrichment(pathways[[changing_pathways$pathway[8]]], changes_lfc, ticksSize = 1) + labs(title=changing_pathways$pathway[8])

```


# Pathway 9: WP Nuclear Receptors Metapathway

EXPLANATION: 
0 hits for "Nuclear Receptors Metapathway cachexia" search...

Target of NRF2 which is associated with ferroptosis resistance in non-small cell lung carcinoma (https://www.nature.com/articles/s41419-023-06070-x)

"because miRNA genes are transcribed by RNA polymerase II, their expression can be also regulated by nuclear receptors [21]. For example an aberrant expression of nuclear receptors such as estrogen receptor α and androgen receptor in cancer is involved in the deregulation of numerous miRNAs" (https://academic.oup.com/jid/article/212/suppl_1/S74/1052160)
```{r}

plotEnrichment(pathways[[changing_pathways$pathway[9]]], changes_lfc, ticksSize = 1) + labs(title=changing_pathways$pathway[9])

```


```{r setting up ClusterProfiler/GO}
## loading in libraries
library(pheatmap)
library("grid")
# library("gridExtra")
library(cluster)
library(org.Hs.eg.db)
library(clusterProfiler)
library(biomaRt)
library(tidyr)



```


```{r ClusterProfiler / GO}

# Converting gene symbol to ensembl ID
ensembl <- useEnsembl(biomart="ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl")
ensembl_query <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "description"),
      filters = "hgnc_symbol",
      values = data.norm.filtered$genes$Symbol,
      mart = ensembl)

length(unique(data.norm.filtered$genes$Symbol)) # 14428 genes
length(unique(ensembl_query$hgnc_symbol)) # 9983 
length(unique(ensembl_query$hgnc_symbol)) / length(unique(data.norm.filtered$genes$Symbol)) * 100 #only 70% of genes map to a ensembl ID

de_down_results$ensembl_id = rep(NA, nrow(de_down_results))
for (i in seq(nrow(de_down_results))){
  if (de_down_results$Symbol[i] %in% ensembl_query$hgnc_symbol) {
    de_down_results$ensembl_id[i] = paste(ensembl_query$ensembl_gene_id[ensembl_query$hgnc_symbol == de_down_results$Symbol[i]], sep = "/", collapse = "/")
  } else {
    de_down_results$ensembl_id[i] = NA
  }
}
de_down_results = separate_longer_delim(de_down_results, "ensembl_id", delim = "/")


de_up_results$ensembl_id = rep(NA, nrow(de_up_results))
for (i in seq(nrow(de_up_results))){
  if (de_up_results$Symbol[i] %in% ensembl_query$hgnc_symbol) {
    de_up_results$ensembl_id[i] = paste(ensembl_query$ensembl_gene_id[ensembl_query$hgnc_symbol == de_up_results$Symbol[i]], sep = "/", collapse = "/")
  } else {
    de_up_results$ensembl_id[i] = NA
  }
  
}
de_up_results = separate_longer_delim(de_up_results, "ensembl_id", delim = "/")



#GO ontology: does hypergeometric test
ego_up <- enrichGO(gene         = de_up_results$Symbol, # de_up_results$ensembl_id, #upregulated genes
                universe      = data.norm.filtered$genes$Symbol, # ensembl_query$ensembl_gene_id, #all expressed genes
                OrgDb         = org.Hs.eg.db, #human
                keyType       = 'SYMBOL', #'ENSEMBL', 
                ont           = "ALL",
                pAdjustMethod = "BH", #correct for multi testing
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.1)  #FDR
head(ego_up)
ego_up_df = as.data.frame(ego_up)
dotplot(ego_up)

#GO ontology: does hypergeometric test
ego_down <- enrichGO(gene     = de_down_results$Symbol, #de_down_results$ensembl_id, #upregulated genes
                universe      = data.norm.filtered$genes$Symbol, #ensembl_query$ensembl_gene_id, #all expressed genes
                OrgDb         = org.Hs.eg.db, #human
                keyType       = 'SYMBOL', #'ENSEMBL', 
                ont           = "ALL",
                pAdjustMethod = "BH", #correct for multi testing
                pvalueCutoff  = 1,
                qvalueCutoff  = 0.1)  #FDR
head(ego_down)
ego_down_df = as.data.frame(ego_down)
dotplot(ego_down)


```





How do we get this to be better? (TO DO by Sept 28)
# use this to filter de_mirna_up_targets to only show genes expresed in muscle
# [SKIP] try running this again without RXF 2
# [DONE]filter results df to remove the ones without genes 
# [done I THINK?] filter to use only one probe per gene --> which probe to use? --> pick probe with highest interquartile range (most variable)

TO DO by Oct 5:
# Take ALL EXPRESSED genes and put them in GSEA --> do not use GO and clusterprofiler
# [DONE] remove the probes without expression --> detectionpvalues() --> figure out propexpr cutoff (? 0.05)
# [DONE] try all 3filter steps with and without RXF 2

TO DO by Oct
# [DONE] Look at the different pathways with padj < 0.1 --> vaildate this dataset, is this useful?
  # [DONE] Pathway 1: Hallmark Epithelial Mesenchymal Transition 
  # [DONE] Pathway 2: Hallmark Hypoxia
  # [DONE] Pathway 3: Hallmark TNFA Signaling via NFKB
  # [DONE] Pathway 4: KEGG Pathways in Cancer
  # [DONE] Pathway 5: NABA Matrisome
  # [DONE] Pathway 6: NABA Matrisome Associated
  # [DONE] Pathway 7: PID FRA Pathway
  # [DONE] Pathway 8: WP Adipogenesis
  # [DONE] Pathway 9: WP Nuclear Receptors Metapathway
# run clusterprofiler on up and downregulated genes separately
# [DONE] filter miRNA de with genes de
  # enrichment analysis, 
  # fold changes of genes, - +/- 2 fold 
  # context score filters -0.4 and below
  
TO DO by Nov. 2
# [DONE] Run GSEA with set.seed and eps again --> see what happens to the pathways 
# [DONE] Run clusterprofiler and GO on up and down genes as two separate clusters
# [DONE] make volcano for SKR vs RXF in microarray
# [DONE] Run clusterprofiler and GO on up and down miRNA targets as two separate clusters
  # try bioMart for the conversion between ENSEMBL and Symbol
  # narrow down to those working in muscles
# [DONE] verify that upregulated miRNA gene targets are going down in microarray data (and vice versa)
  # upregulated miRNA: 177 / 340
  # downregulated miRNA: 328 / 719
# [DONE] get percentage of targets per miRNA 
# make those three plots prof was talking about -- for ranking miRNA targets

TO DO Nov 2.
# look at neurons in terms of cachexia
# [DONE] add a cut off log2fold change of the gene expression target (1.5)
# [DONE] make table of genes with excel  --> with ranks for context score, fold change of target, fold change of miRNA, rank based on expression; make filters for description
  # [DONE] prioritizing miRNA-target pairs --> get top X pairs by each parameter and find overlaps; keep some certain genes 
    based on biology (e.g., SIX1, RARA, RARB, SERPINH1, TSKU)
# [PENDING] make presentation 
 # DE
 # microarray
 # filtered miRNA
 # proposal
# [DONE] send prof table of miRNAs
# [PENDING] fix plots
  # ASK PROF about plots for downregulated miRNA gene targets


TO DO 
# reading stuff
# [PENDING] full excel sheet with ALL miRNAs with filters on interesting terms
  # sheet 1: DE results of miRNAs (LFC, p-value, UP/DOWN)
  # Sheet 2: DE results of microarray (LFC, p-value, UP/DOWN)
  # Sheet 3: all miRNAs - microarray (without ANY filters)
# make a presentation
  # results 
# [DONE] fix figure, dont have boxplot plot the outlier
# [DONE] make general plot of upregulated, downregulated
# [PENDING] ZNF609


```{r}
sessionInfo()
```